File: qbopt/optimize/ivshare.py
from dataclasses import replace

from qbopt.model import mir
from qbopt.analysis import loops
from qbopt.analysis import consts
from qbopt.analysis import induction


def shared(body: mir.MirBody) -> mir.MirBody:
    from qbopt.optimize import transform
    from qbopt.optimize.strength import _made

    definitions = {value.id: op for block in body.blocks for op in block.ops for value in op.defines}
    required = transform.halves(body)
    for loop in loops.loops(body.blocks, body.entry):
        counters = induction.basics(body, loop)
        header = next(block for block in body.blocks if block.at == loop.header)
        for derived in counters.values():
            twin = _twin(counters, derived, header, required)
            if twin is not None:
                return _replacing(body, header, derived, twin, derived.start.width)
            found = _offset(counters, derived, definitions)
            if found is None:
                continue
            base, offset, seed = found
            width = derived.start.width
            phi = next(one for one in header.phis if one.result.id == derived.value)
            upper = None
            if width < 4 and (phi.result, transform.HIGH) in required:
                carried = []
                for incoming in phi.incoming.values():
                    producer = definitions.get(incoming.id)
                    if (
                        producer is None
                        or producer.results != (mir.Held(incoming, width),)
                        or len(producer.merges) != 1
                    ):
                        break
                    carried.append(next(iter(producer.merges)))
                else:
                    if len(set(carried)) == 1:
                        upper = carried[0]
                if upper is None or upper.id not in induction.invariant(body, set(loop.body)):
                    continue
            root = next(one.result for one in header.phis if one.result.id == base.value)
            operation = _made(
                mir.Kind.ADD, "add", phi.result, (mir.Held(root, width), offset), header.at, seed or header.ops[0]
            )
            if upper is not None:
                operation = replace(operation, merges={upper: phi.result}, uses=(*operation.uses, upper))
            changed = replace(
                header, phis=tuple(one for one in header.phis if one is not phi), ops=(operation, *header.ops)
            )
            return replace(body, blocks=tuple(changed if block is header else block for block in body.blocks))
    return body


def _offset(
    counters: dict[int, induction.Affine], derived: induction.Affine, definitions: dict[int, mir.Op]
) -> tuple[induction.Affine, mir.Const, mir.Op | None] | None:
    """Another counter stepping as `derived` does a constant distance behind it, the distance, and a seed.

    Two starts are that far apart when both are one root plus a constant:
    `add source,c` over the other's start, or strength reduction's `a[i].x`
    and `a[i].y`, starting 4 apart from the same or no root. Otherwise the
    canonical counter is the lower id, so the pair converges.
    """
    if not isinstance(derived.step, mir.Const) or not isinstance(derived.start, (mir.Held, mir.Const)):
        return None
    width = derived.start.width
    root, at = _anchor(derived.start, definitions, width)
    seed = definitions.get(derived.start.value.id) if isinstance(derived.start, mir.Held) else None
    direct = seed.args if seed is not None and seed.kind is mir.Kind.ADD else ()
    for one in sorted(counters.values(), key=lambda one: (one.start not in direct, one.value)):
        if one.value == derived.value or one.step != derived.step or not isinstance(one.start, (mir.Held, mir.Const)):
            continue
        if one.start not in direct and one.value > derived.value:
            continue
        if one.start.width != width or _anchor(one.start, definitions, width)[0] != root:
            continue
        distance = at - _anchor(one.start, definitions, width)[1]
        return one, mir.Const(consts.masked(distance, width), width), seed
    return None


def _anchor(arg: mir.Held | mir.Const, definitions: dict[int, mir.Op], width: int) -> tuple[mir.Value | None, int]:
    """`arg` as a root value plus a constant, through copies and constant adds; a constant has no root."""
    offset = 0
    while isinstance(arg, mir.Held) and arg.width == width:
        op = definitions.get(arg.value.id)
        if op is None or op.loads or op.stores or op.barrier or op.merges or op.results != (arg,):
            break
        if op.kind is mir.Kind.COPY and len(op.args) == 1 and isinstance(op.args[0], (mir.Held, mir.Const)):
            arg = op.args[0]
        elif op.kind is mir.Kind.ADD and len(op.args) == 2 and sum(isinstance(one, mir.Const) for one in op.args) == 1:
            constant, arg = sorted(op.args, key=lambda one: not isinstance(one, mir.Const))
            offset += constant.n
        else:
            break
    if isinstance(arg, mir.Const):
        return None, consts.masked(arg.n + offset, width)
    return arg.value, offset


def _twin(counters, derived, header, required):
    """Another counter of this loop that advances identically, or None.

    The offset case below is the general one and this is its degenerate
    form: two recurrences with the same start and the same step are the
    same value, and `offset` would be zero. It is not reachable through
    that path -- it looks for a start defined by `add source,const`, and
    two identical counters share a start rather than computing one from the
    other -- and nothing else merges them, because a phi is not one of the
    computations `transform`'s value numbering considers.

    BC writes one for every use of a subscript, so segld's inner loop
    advanced two offsets into `a` that were always equal. The cost is not
    only the extra `add`: `mir.same_bytes` keys a reference on its base
    *value*, so `a(i)` stored through one and read back through the other
    were two different cells and the reload could not be forwarded.

    Canonical is the lower id, so eliminating the higher one converges
    rather than swapping the pair back and forth across rounds.
    """
    from qbopt.optimize import transform

    if any(one.result.id == derived.value and (one.result, transform.HIGH) in required for one in header.phis):
        return None  # a long pair's high half is tied to this phi; the offset path owns that
    return next(
        (
            one
            for one in counters.values()
            if one.value < derived.value and one.start == derived.start and one.step == derived.step
        ),
        None,
    )


def _replacing(body, header, derived, twin, width):
    """`body` with `derived`'s phi replaced by a copy of `twin`'s."""
    from qbopt.optimize.strength import _made

    phi = next(one for one in header.phis if one.result.id == derived.value)
    root = next(one.result for one in header.phis if one.result.id == twin.value)
    operation = _made(mir.Kind.COPY, "mov", phi.result, (mir.Held(root, width),), header.at, header.ops[0])
    changed = replace(header, phis=tuple(one for one in header.phis if one is not phi), ops=(operation, *header.ops))
    return replace(body, blocks=tuple(changed if block is header else block for block in body.blocks))

File: crates/llrm-core/src/optimize/ivshare.rs
//! Direct port of `qbopt/optimize/ivshare.py`.
//!
//! Tests: every test in `tests/test_ivshare.py` is skipped; its `culling`
//! fixture spies on `strength.reduced`, and errors in Python at this commit.

use std::rc::Rc;
use std::collections::{BTreeMap, BTreeSet};

use crate::analysis::consts::masked;
use crate::analysis::induction::{self, Affine, AffineOperand};
use crate::analysis::loops;
use crate::model::mir::{Arg, Const, Held, Kind, MirBody, MirBlock, Op, OrderedMap, Value};
use crate::optimize::strength::_made;
use crate::optimize::transform;

/// Direct port of `qbopt/optimize/ivshare.py:shared`.
pub fn shared(body: &Rc<MirBody>) -> Rc<MirBody> {
    let mut definitions: BTreeMap<u32, &Op> = BTreeMap::new();
    let mut made: BTreeMap<Value, &Op> = BTreeMap::new();
    for block in &body.blocks {
        for op in &block.ops {
            for value in &op.defines {
                definitions.insert(value.id, op);
                made.insert(*value, op);
            }
        }
    }
    let required = transform::halves(body);
    for loop_ in loops::loops(&body.blocks, Some(body.entry)) {
        let counters = induction::basics(body, &loop_);
        let header_index = body
            .blocks
            .iter()
            .position(|block| block.at == loop_.header)
            .expect("loop header is a block");
        let header = &body.blocks[header_index];
        for derived in counters.values() {
            let twin = _twin(&counters, derived, header, &required);
            if let Some(twin) = twin {
                return Rc::new(_replacing(body, header_index, derived, twin, derived.start.width()));
            }
            let Some((base, offset, seed)) = _offset(&counters, derived, &definitions, &made) else {
                continue;
            };
            let width = derived.start.width();
            let phi_index = header
                .phis
                .iter()
                .position(|one| one.result.id == derived.value)
                .expect("counter has a header phi");
            let phi = &header.phis[phi_index];
            let mut upper: Option<Value> = None;
            if width < 4 && required.contains(&(phi.result, transform::HIGH)) {
                let mut carried = Vec::new();
                let mut broke = false;
                for incoming in phi.incoming.values() {
                    let producer = definitions.get(&incoming.id).copied();
                    let Some(producer) = producer.filter(|producer| {
                        producer.results
                            == vec![Arg::Held(Held {
                                value: *incoming,
                                width,
                            })]
                            && producer.merges.len() == 1
                    }) else {
                        broke = true;
                        break;
                    };
                    carried.push(*producer.merges.keys().next().expect("one merge"));
                }
                if !broke && carried.iter().collect::<BTreeSet<_>>().len() == 1 {
                    upper = Some(carried[0]);
                }
                match upper {
                    Some(value) if induction::invariant(body, &loop_.body).contains(&value.id) => {}
                    _ => continue,
                }
            }
            let root = header
                .phis
                .iter()
                .find(|one| one.result.id == base.value)
                .expect("base counter has a header phi")
                .result;
            let mut operation = _made(
                Kind::Add,
                "add",
                phi.result,
                vec![Arg::Held(Held { value: root, width }), Arg::Const(offset)],
                header.at,
                seed.unwrap_or(&header.ops[0]),
            );
            if let Some(upper) = upper {
                let mut merges = OrderedMap::new();
                merges.insert(upper, phi.result);
                operation.merges = merges;
                operation.uses.push(upper);
            }
            let mut changed = header.clone();
            changed.phis.remove(phi_index);
            changed.ops.insert(0, operation);
            let mut result = MirBody::clone(body);
            result.blocks[header_index] = changed;
            return Rc::new(result);
        }
    }
    body.clone()
}

/// Another counter stepping as `derived` does a constant distance behind it, the distance, and a seed.
///
/// Two starts are that far apart when `induction::distance` says so:
/// `add source,c` over the other's start, or strength reduction's `a[i].x`
/// and `a[i].y`, starting 4 apart. Otherwise the canonical counter is the
/// lower id, so the pair converges.
fn _offset<'a, 'b>(
    counters: &'a OrderedMap<u32, Affine>,
    derived: &Affine,
    definitions: &BTreeMap<u32, &'b Op>,
    made: &BTreeMap<Value, &'b Op>,
) -> Option<(&'a Affine, Const, Option<&'b Op>)> {
    if !matches!(derived.step, AffineOperand::Const(_)) {
        return None;
    }
    let width = derived.start.width();
    let seed = match &derived.start {
        AffineOperand::Held(start) => definitions.get(&start.value.id).copied(),
        AffineOperand::Const(_) => None,
    };
    let direct: &[Arg] = match seed {
        Some(seed) if seed.kind == Kind::Add => &seed.args,
        _ => &[],
    };
    let mut ordered: Vec<&Affine> = counters.values().collect();
    ordered.sort_by_key(|one| (!direct.contains(&one.start.as_arg()), one.value));
    for one in ordered {
        if one.value == derived.value || one.step != derived.step {
            continue;
        }
        if !direct.contains(&one.start.as_arg()) && one.value > derived.value {
            continue;
        }
        if one.start.width() != width {
            continue;
        }
        let Some(distance) = induction::distance(&derived.start.as_arg(), &one.start.as_arg(), made, width) else {
            continue;
        };
        return Some((one, Const::new(distance, width), seed));
    }
    None
}

/// Another counter of this loop that advances identically, or None.
///
/// Direct port of `qbopt/optimize/ivshare.py:_twin`.
fn _twin<'a>(
    counters: &'a OrderedMap<u32, Affine>,
    derived: &Affine,
    header: &MirBlock,
    required: &BTreeSet<(Value, u8)>,
) -> Option<&'a Affine> {
    if header
        .phis
        .iter()
        .any(|one| one.result.id == derived.value && required.contains(&(one.result, transform::HIGH)))
    {
        return None; // a long pair's high half is tied to this phi; the offset path owns that
    }
    counters
        .values()
        .find(|one| one.value < derived.value && one.start == derived.start && one.step == derived.step)
}

/// `body` with `derived`'s phi replaced by a copy of `twin`'s.
///
/// Direct port of `qbopt/optimize/ivshare.py:_replacing`.  `header` is the
/// block's index: Python compares the block by identity.
fn _replacing(body: &MirBody, header: usize, derived: &Affine, twin: &Affine, width: u32) -> MirBody {
    let block = &body.blocks[header];
    let phi_index = block
        .phis
        .iter()
        .position(|one| one.result.id == derived.value)
        .expect("counter has a header phi");
    let phi = &block.phis[phi_index];
    let root = block
        .phis
        .iter()
        .find(|one| one.result.id == twin.value)
        .expect("twin counter has a header phi")
        .result;
    let operation = _made(
        Kind::Copy,
        "mov",
        phi.result,
        vec![Arg::Held(Held { value: root, width })],
        block.at,
        &block.ops[0],
    );
    let mut changed = block.clone();
    changed.phis.remove(phi_index);
    changed.ops.insert(0, operation);
    let mut result = body.clone();
    result.blocks[header] = changed;
    result
}
